### PR TITLE
setup: Convert setup script to python

### PR DESCRIPTION
**Pull request by lsgunth**
Convert the setup script to python and add the ability to start
with loop disks or disks specified on the command line.

Martin: you can probably make use of this code in your work as well.

### CODE REVIEW

File: md.py [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+        self.wait()
+        self.stop()
+
+        if disks is None and ram_disks is None and loop_disks is None:
+            raise MDInvalidArgumentError("No disks specified for an array")
+
+        if ram_disks is not None:
+            if disks:
+                raise MDInvalidArgumentError("Must not specify both disks and ram_disks")
+            subprocess.check_call(["modprobe", "brd"])
```
Comment (iomartin) [inline md.py]: This should be `modprobe brd rd_nr=count(ram_disks)` so that we create the right number of ram disks

File: md.py [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+            mdadm_args.append("--force")
+        if run:
+            mdadm_args.append("--run")
+        if quiet:
+            mdadm_args.append("--quiet")
+        if journal is not None:
+            mdadm_args += ["--write-journal", journal]
+        if size is not None:
+            mdadm_args += ["--size", str(size)]
+
```
Comment (iomartin) [inline md.py:140]: nit: two empty lines

File: md.py [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+
+        print(mdadm_args + disks)
```
Comment (iomartin) [inline md.py]: Is this a debug print that was left behind?

File: setup [hunk 4]
```diff
@@ -1,72 +1,119 @@
...
+
+    inst = md.MDInstance()
+    inst.setup(level=args.level,
+               ram_disks=args.disks,
+               loop_disks = args.loop_disks,
+               size=args.mdsize,
+               chunk_size=args.chunk_size,
+               assume_clean=args.assume_clean,
+               force=args.force,
+               run=args.run or args.zero_first,
```
Comment (iomartin) [inline setup]: args.run and args.zero_first are only used here, can't they just be aliases to the same argument?

File: setup [hunk 5]
```diff
@@ -1,72 +1,119 @@
...
+    if not args.devs and not args.loop_disks and not args.disks:
+        args.disks = 3
+
+    if os.geteuid() != 0:
+        sys.exit("Must run as root")
+
+    inst = md.MDInstance()
+    inst.setup(level=args.level,
+               ram_disks=args.disks,
+               loop_disks = args.loop_disks,
```
Comment (iomartin) [inline setup]: nit: extra space around =

File: md.py [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+            open(backing, "w").close()
+            os.truncate(backing, size)
+            subprocess.check_call(["losetup", dev, backing])
+            ret.append(dev)
+
+        return ret
+
+    def setup(self, level=5, disks=None, ram_disks=None, loop_disks=None,
+              size=None, chunk_size=64 << 10, assume_clean=True, force=True,
+              run=False, policy="resync", journal=False, quiet=False,
```
Comment (iomartin) [inline md.py]: Did you mean `journal=None`?

This is later used as

```python
if journal is not None:
    mdadm_args += ["--write-journal", journal]
```

File: md.py [hunk 7]
```diff
@@ -0,0 +1,91 @@
...
+
+            disks = self._create_loop_disks(loop_disks, size)
+            size = None
+
+        mdadm_args = ["mdadm", "--create", self._md_dev,
+                      "--level", str(level),
+                      "--chunk", str(chunk_size),
+                      "--raid-devices", str(len(disks))]
+
+        if policy == "bitmap":
```
Comment (iomartin) [inline md.py:126]: How about the other policies?

File: setup [hunk 8]
```diff
@@ -1,72 +1,119 @@
...
+                        help="group thread count for array")
+    parser.add_argument("--cache-size", default=8192, type=int,
+                        help="cache size")
+    parser.add_argument("--journal", help="journal type")
+    parser.add_argument("--size", "--mdsize", dest="mdsize",
+                        type=suffix_parse, help="array size")
+
+    args = parser.parse_args()
+
+    if args.devs:
```
Comment (iomartin) [inline setup]: Isn't all of that guaranteed by the `parser.add_mutually_exclusive_group()`?

File: md.py [hunk 9]
```diff
@@ -0,0 +1,91 @@
...
+
+        self.wait()
+        self.stop()
+
+        if disks is None and ram_disks is None and loop_disks is None:
+            raise MDInvalidArgumentError("No disks specified for an array")
+
+        if ram_disks is not None:
+            if disks:
+                raise MDInvalidArgumentError("Must not specify both disks and ram_disks")
```
Comment (iomartin) [inline md.py:107]: Don't we need to check that ram_disks is also None?

File: setup [hunk 10]
```diff
@@ -1,72 +1,119 @@
...
+                        help="consistency policy")
+    parser.add_argument("--quiet", action="store_true",
+                        help="be quiet")
+    parser.add_argument("--thread-cnt", default=4, type=int,
+                        help="group thread count for array")
+    parser.add_argument("--cache-size", default=8192, type=int,
+                        help="cache size")
+    parser.add_argument("--journal", help="journal type")
+    parser.add_argument("--size", "--mdsize", dest="mdsize",
+                        type=suffix_parse, help="array size")
```
Comment (iomartin) [inline setup]: This should say that if --size is used with --loop-devices, then --size is the size of the loop devices

File: md.py (hunk 1 above)
Comment (iomartin) [reply md.py]: Maybe use `rd_size=<bytes>` as well so that the disks aren't so tiny (the default is 4K)?

Comment (lsgunth) [reply md.py]: The count won't work as some tests expect to grow more disks. Also the module may already be loaded so these parameters might be ignored... But I can add a "default" rd_size.

File: setup (hunk 4 above)
Comment (lsgunth) [reply setup]: "run" is a better name which matches mdadm but I've left zero_first for backwards compatibility with some of the other scripts.

File: setup (hunk 8 above)
Comment (lsgunth) [reply setup]: No... because of the Environment variable parsing the defaults can bypass that.

File: md.py (hunk 9 above)
Comment (lsgunth) [reply md.py:107]: What?

Comment (iomartin) [reply md.py:107]: Ops, I meant loop_disks

File: setup (hunk 4 above)
Comment (iomartin) [reply setup]: But can't it be `parser.add_argument("--zero-first", "--run")`?

File: setup (hunk 10 above)
Comment (lsgunth) [reply setup]: Well it's actually the amount of bytes taken by each disk in both situations; so I've reworded it.

File: md.py [hunk 11]
```diff
@@ -0,0 +1,204 @@
...
+import re
+import subprocess
+
+class EnvironmentArgumentParser(argparse.ArgumentParser):
+    class _CustomHelpFormatter(argparse.ArgumentDefaultsHelpFormatter):
+        def _get_help_string(self, action):
+            help = super()._get_help_string(action)
+            if action.dest != 'help':
+                help += ' [env: {}]'.format(action.dest.upper())
+                return help
```
Comment (iomartin) [inline md.py]: The `return help` needs to be outside of the `if`

File: md.py (hunk 9 above)
Comment (lsgunth) [reply md.py:107]: disks gets set in loop disk so it doesn't need to explicitly also check ram_disks

File: setup (hunk 4 above)
Comment (lsgunth) [reply setup]: The environement parser thingy only takes the dest argument as where to check the environment. So I want to check both environment variables.

File: md.py [hunk 12]
```diff
@@ -0,0 +1,204 @@
...
+            return help
+
+    def __init__(self, *, formatter_class=_CustomHelpFormatter,
+                 **kwargs):
+        super().__init__(formatter_class=formatter_class, **kwargs)
+
+    def to_bool(self, x):
+        if isinstance(x, bool):
+            return x
+        if isinstance(x, str):
```
Comment (dlsloan) [inline md.py:24]: Can we add 1 for true too, that's fairly common too

File: md.py (hunk 9 above)
Comment (dlsloan) [reply md.py:107]: But shouldn't it validate the args before calling modprobe? seems weird to get part-way through init and then get an invalid argument exception

Comment (lsgunth) [reply md.py:107]: Honestly, don't care if someone inadvertantly modprobes brd with a poorly constructed set of arguments.

File: md.py [hunk 13]
```diff
@@ -5,6 +5,8 @@
 import pathlib
 import re
 import subprocess
+import sys
```
Comment (iomartin) [inline md.py]: In the commit message:
s/then immediately stopping the arary./then immediately creating the array

**Review by iomartin — APPROVED** [review]
Just one small thing on a commit message but otherwise looks good to me

File: md.py (hunk 13 above)
Comment (lsgunth) [reply md.py]: Actually hmmm. I'm investigating the possibility that the need for a delay here has been caused by a bug in a recent patch. I'm also failing an mdadm test all of a sudden.

**Review by dlsloan — APPROVED** [review]

**Comment by lsgunth** [comment]
Dropped the patch that was actually caused by a kernel regression so this can be merged.

**Pull request merged**